File: visual_feedback/patch_vision/include/patch_vision/matching/match_io.cpp
/*
 * =====================================================================================
 *
 *       Filename:  match_io.cpp
 *
 *    Description:  
 *
 *        Version:  1.0
 *        Created:  07/06/2011 07:53:11 AM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Stephen Miller (stephen), sdmiller@eecs.berkeley.edu
 *        Company:  UC Berkeley
 *
 * =====================================================================================
 */
#include <patch_vision/matching/match_io.h>

namespace patch_vision{

using std::ofstream;
using std::ifstream;
using std::endl;
using std::cout;

MatchSet :: MatchSet( ){
    clear();
}


void MatchSet :: add_match( float x_c, float y_c, float x_r, float y_r, float strength, int ref_idx ){
    add_match( PointT(x_c,y_c), PointT(x_r,y_r), strength );
}
void MatchSet :: add_match( const PointT point_c, const PointT point_r, float strength, int ref_idx ){
    Match match;
    match.point_c = PointT(point_c);
    match.point_r = PointT(point_r);
    match.strength = strength;
    match.ref_idx = ref_idx;
    _matches.push_back(match);
}

int  MatchSet :: num_matches( ) const{
    return _matches.size();
}

void MatchSet :: get_matches( vector<Match> &matches ) const{
    for (unsigned int i = 0; i < _matches.size(); i++ ){
        Match match;
        match.point_c = PointT(_matches[i].point_c);
        match.point_r = PointT(_matches[i].point_r);
        match.strength = _matches[i].strength;
        match.ref_idx = _matches[i].ref_idx;
        matches.push_back(match);
    }
}

void MatchSet :: save_to_file( string filename ) const{
    ofstream output_file( filename.c_str() );
    output_file << _matches.size() << endl;
    for( size_t i = 0; i < _matches.size(); i++ ){
        PointT pt_c = _matches[i].point_c;
        PointT pt_r = _matches[i].point_r;
        float strength = _matches[i].strength;
        int ref_idx = _matches[i].ref_idx;
        output_file <<  pt_c.first  << " " << pt_c.second << "\t"
                    <<  pt_r.first  << " " << pt_r.second << "\t"
                    <<  strength << " " << ref_idx << endl;
    }
    output_file.close();
}
void MatchSet :: read_from_file( string filename ){
    ifstream input_file( filename.c_str() );
    size_t num_matches;
    input_file >> num_matches;
    for( size_t i = 0; i < num_matches; i++ ){
        PointT pt_c;
        PointT pt_r;
        float strength;
        int ref_idx;
        input_file  >>  pt_c.first  >> pt_c.second 
                    >>  pt_r.first  >> pt_r.second
                    >>  strength    >> ref_idx;
        add_match( pt_c, pt_r, strength, ref_idx );
    }
    input_file.close();

}

void MatchSet :: clear(){
    _matches.clear();
}

};

File: visual_feedback/patch_vision/src/patch_vision/matching/match_io.py
#!/usr/bin/env python

class Match:
    def __init__(self, compare_pt, reference_pt, strength, ref_idx=-1):
        self.compare_pt = compare_pt
        self.reference_pt = reference_pt
        self.strength = strength
        self.ref_idx = ref_idx

class MatchSet:
    def __init__(self, compare_image_path, reference_image_path):
        self.compare_image_path = compare_image_path
        self.reference_image_path = reference_image_path
        self.matches = []

    def add_match(self, compare_pt, reference_pt, strength, ref_idx = -1):
        self.matches.append( Match( compare_pt, reference_pt, strength, ref_idx) )

    def add_match_obj(self, match):
        self.matches.append( match )

    def get_matches(self):
        return self.matches

    def save_to_file(self, filename):
        f = open(filename,'w')
        f.write("%d\n"%len(self.get_matches()))
        for match in self.get_matches():
            f.write("%f %f"%(match.compare_pt[0], match.compare_pt[1]))
            f.write("\t")
            f.write("%f %f"%(match.reference_pt[0], match.reference_pt[1]))
            f.write("\t")
            f.write("%f"%match.strength)
            f.write("\t")
            f.write("%d"%match.ref_idx)
            f.write("\n")
        f.close()

    def read_from_file(self, filename):
        f = open(filename, 'r')
        for i,ln in enumerate(f.readlines()):
            if i == 0:
                continue
            vals = [float(v) for v in ln.split()]
            compare_pt = (vals[0], vals[1])
            reference_pt = (vals[2], vals[3])
            strength = vals[4]
            ref_idx = -1 if len(vals) < 6 else int(ln.split()[5])
            self.matches.append( Match( compare_pt, reference_pt, strength, ref_idx ) )
        f.close()
        
        
